web/client/utils/StateUtils.js

```diff
@@ -136,6 +117,100 @@ export const getState = (name) => {
+};
+
+
+export const createStoreManager = (initialReducers, initialEpics) => {
+    // Create an object which maps keys to reducers
+    const reducers = {...initialReducers};
+    const epics = {...initialEpics};
+
+    // Create the initial combinedReducer
+    let combinedReducer = combineReducers(reducers);
+
+    const epic$ = new BehaviorSubject(combineEpics(...wrapEpics(epics)));
+
+    // An array which is used to delete state keys when reducers are removed
+    let keysToRemove = [];
+
+    let muteState = {};
+
+    return {
+        getReducerMap: () => reducers,
+
+        // The root reducer function exposed by this object
+        // This will be passed to the store
+        reduce: (state, action) => {
+            // If any reducers have been removed, clean up their state first
+            if (keysToRemove.length > 0) {
+                // eslint-disable-next-line no-param-reassign
+                state = {...state};
+                for (let key of keysToRemove) {
+                    delete state[key];
+                }
+                keysToRemove = [];
+            }
+
+            // Delegate to the combined reducer
+            return combinedReducer(state, action);
+        },
+
+        // Adds a new reducer with the specified key
+        addReducer: (key, reducer) => {
+            if (!key || reducers[key]) {
+                return;
+            }
+
+            // Add the reducer to the reducer mapping
+            reducers[key] = reducer;
+
+            // Generate a new combined reducer
+            combinedReducer = combineReducers(reducers);
+        },
+
+        // Removes a reducer with the specified key
+        removeReducer: key => {
+            if (!key || !reducers[key]) {
+                return;
+            }
+
+            // Remove it from the reducer mapping
+            delete reducers[key];
+
+            // Add the key to the list of keys to clean up
+            keysToRemove.push(key);
+
+            // Generate a new combined reducer
+            combinedReducer = combineReducers(reducers);
+        },
+        // Adds a new epics set, mutable by the specified key
+        addEpics: (key, epicsList) => {
+            if (Object.keys(epicsList).length) {
+                muteState[key] = new Subject();
+                const isolatedEpics = isolateEpics(epicsList, muteState[key].asObservable());
+                wrapEpics(isolatedEpics).forEach(epic => epic$.next(epic));
+            }
+        },
+        // Mute epics set with a specified key
+        muteEpics: (key) => {
+            if (typeof muteState[`${key}Plugin`] !== 'undefined') {
+                muteState[`${key}Plugin`].next(false);
+            }
+        },
+        unmuteEpics: (key) => {
+            if (typeof muteState[`${key}Plugin`] !== 'undefined') {
+                muteState[`${key}Plugin`].next(true);
+            }
+        },
```

Comment:
Just tested the mute and unmute functionality for epics and it is not working and the reason seems related to the key used in the muteState global object cache because the key sometimes contain the `Plugin` suffix and sometimes not. The only solution I came up is to trim the end always but maybe there is a better one:

```js
const pluginName = trimEnd(key, 'Plugin');
muteState[pluginName]
```

With the correct key it seems to work for common epic such as

```js
const msTestEpic = (action$) => {
    return action$.ofType('@@router/LOCATION_CHANGE')
        .switchMap(() => {
            console.log('TEST');
            return Observable.empty();
        });
};
```

while epic that has a subprocess that does not depend on the action (eg timer) are not "muted" (probably they are muted but the subprocess does not know if it should stop or not). I don't think these is a common use case but it could happen that you need a polling stream in a extension/plugin

```js
const msTestEpicTimer = (action$) => {
    return action$.ofType('@@router/LOCATION_CHANGE')
        .switchMap(() => {
            return Observable.interval(100)
                .switchMap(() => {
                    console.log('TEST');
                    return Observable.empty();
                });
        });
};
```

Reply:
Here is the working snippet that actually turns off substream whenever isActive$ become false. Basically, code is the same as for the outer wrapper of each isolatedEpic. With the recent commit both modulePlugins epics and appEpics are wrapped wtih isolateEpics, so all of them will have third argument available, even though it will have no effect on appEpics since there is no code to mute them.
```js
export const dummyEpic = (action$, store, isActive$) => action$.ofType(ACTION)
    .switchMap(() => {
        return Rx.Observable.interval(1000)
            .let(semaphore(isActive$.startWith(true)))
            .switchMap(() => {
                console.log('TEST');
                return Rx.Observable.empty();
            });
    });
```